# Python basics

Computational Finance with Python

[Alet Roux](https://www.york.ac.uk/maths/staff/alet-roux/) ([Department
of Mathematics](https://maths.york.ac.uk), University of York)

Click on the following to open this file in Google Colab:

<figure>
<a
href="https://colab.research.google.com/github/aletroux/comp-finance-python/blob/main/demonstrations/D01_Python_basics_slides.ipynb"><img
src="https://colab.research.google.com/assets/colab-badge.svg"
alt="Open In Colab" /></a>
<figcaption>Open In Colab</figcaption>
</figure>

# First things first

## A first example

-   Python code and output:

In [1]:
print("Hello world!")

# Assign a variable and display its value
my_answer = 42
print(my_answer)

Hello world!
42

## Variables

-   To create a variable, give it a name and assign a value using the
    equality sign `=`.

-   Variable names must satisfy the following:

    -   Consist of letters (upper and lower case), numbers, and
        underscores (“`_`”)
    -   Start with a letter or an underscore.
    -   Case-sensitive. For example, `age`, `Age` and `AGE` are three
        different variables.
    -   Python keywords (The Python Software Foundation (2023b)) cannot
        be used.

## Python keywords

-   The Python Software Foundation (2023b):

|            |           |          |            |          |
|------------|-----------|----------|------------|----------|
| `False`    | `None`    | `True`   | `and`      | `as`     |
| `assert`   | `async`   | `await`  | `break`    | `class`  |
| `continue` | `def`     | `del`    | `elif`     | `else`   |
| `except`   | `finally` | `for`    | `from`     | `global` |
| `in`       | `is`      | `lambda` | `nonlocal` | `not`    |
| `return`   | `try`     | `while`  | `with`     | `yield`  |

# Basic data types

-   Python is a *dynamically typed* language, which simplifies
    programming.

-   The built-in function `type` tells us the type of an object if
    needed:

In [2]:
print(type(my_answer))

<class 'int'>

-   Python has a number of basic data types that are useful for our
    work:
    -   Integers
    -   Floating point numbers
    -   Booleans
    -   Strings

## Integers

-   Python integers (`int`) can be arbitrarily large: Python uses as
    many bits as needed.
-   Assign two integer variables and display their values:

In [3]:
a = 5
b = 3
print("a =", a)
print("b =", b)

a = 5
b = 3

------------------------------------------------------------------------

### `int` arithmetic

-   Addition, subtraction and multiplication:

In [4]:
print("a + b =", a + b)
print("a - b =", a - b)
print("a * b =", a * b)

a + b = 8
a - b = 2
a * b = 15

-   Powers:

In [5]:
print("a ** b =", a ** b)
print("a ** (-b) =", a ** (-b))

a ** b = 125
a ** (-b) = 0.008

------------------------------------------------------------------------

### `int` division

In [6]:
print("a / b =", a / b)
print("a // b =", a // b)
print("a % b =", a % b)

a / b = 1.6666666666666667
a // b = 1
a % b = 2

------------------------------------------------------------------------

### Assigning and modifying integer variables

In [7]:
# Assigning a to a new value
a = b + 6
print("a =", a)

a = 9

-   The shorthand operators `+=`, `-=`, `*=`, `/=`, etc, can be used to
    modify a variable directly:

In [8]:
# Subtract 1 from a
a -= 1
print("a =", a)

a = 8

In [9]:
#Now divide the result by 3
a /= 3
print("a =", a)
print("Data type of a is now", type(a)) 

a = 2.6666666666666665
Data type of a is now <class 'float'>

## Floating point numbers

-   `float` is used to represent decimal numbers.

In [10]:
# Assign two floats and display their values
c = 0.35
d = 0.15
print("c =", c)
print("d =", d)

c = 0.35
d = 0.15

-   Python tends to conform to the IEEE 754 double-precision standard,
    which is 15-digit relative accuracy (IEEE (2008)).
-   Higher accuracy can be achieved with the `decimal` module (The
    Python Software Foundation (2023a)).

------------------------------------------------------------------------

### Floating number arithmetic

In [11]:
c = 0.35
d = 0.15
print("c + d =", c + d)
print("c - d =", c - d)
print("c * d =", c * d)
print("c / d =", c / d)
print("c ** d =", c ** d)

c + d = 0.5
c - d = 0.19999999999999998
c * d = 0.0525
c / d = 2.3333333333333335
c ** d = 0.8542996071500324

------------------------------------------------------------------------

### `float` assigning and modification

-   `float` variables are assigned with `=`, just like integers.
-   The value of a `float` variable can be changed after its initial
    definition.
-   The shorthand operators `+=`, `-=`, `*=`, `/=`, etc, can be used for
    compactness.

In [12]:
# calculate 1.5^2.4
f = 1.5 ** 2.4

# then add 10
f += 10

print ("f =",f)

f = 12.646177800680515

------------------------------------------------------------------------

### Rounding

-   Use the built-in function `round` for rounding.
-   Rounding to fixed number of decimal places:

In [13]:
c = 0.35
d = 0.15
print(c/d, "rounded to 3 d.p. is", round(c/d, 3) )
print(c-d, "rounded to 1 d.p. is", round(c-d, 1) )

2.3333333333333335 rounded to 3 d.p. is 2.333
0.19999999999999998 rounded to 1 d.p. is 0.2

-   Rounding can yield unexpected results:

In [14]:
print ("Nearest int to", 0.5, "is", round(0.5))
print ("Nearest int to", 1.5, "is", round(1.5))
print ("Data type of rounded num is", type(round(1.5)))

Nearest int to 0.5 is 0
Nearest int to 1.5 is 2
Data type of rounded num is <class 'int'>

## Boolean variables

-   Boolean (`bool`) variables take two values: `True` or `False`.
-   Useful for logical expressions, such as comparisons.

In [15]:
a = 2.66675
d = 0.15

# Less-than and greater than comparison
print("Is a strictly less than d?", a < d)
print("Is a greater than or equal to d?", a >= d)

Is a strictly less than d? False
Is a greater than or equal to d? True

In [16]:
# Equality comparison
print("Is a equal to d?", a == d)
print("Is a not equal to d?", a != d)

Is a equal to d? False
Is a not equal to d? True

------------------------------------------------------------------------

### Logical operators

| Priority | Operator | Definition |
|:-----------------:|:-----------------:|------------------------------------|
| 1 | `not x` | If `x` is `True`, then `False`, otherwise `True`. |
| 2 | `x and y` | `True` if both `x` and `y` are `True`, otherwise `False`. |
| 3 | `x or y` | `True` if at least one of `x` and `y` are `True`, otherwise `False`. |

-   These operators evaluate the first expression (`x`) and depending on
    its value, may or may not evaluate the second (`y`).
-   Logical operators have lower priority than other operators. For
    example, `not a == b` is interpreted as `not (a == b)`. It’s a good
    idea to use brackets regardless for clarity.

------------------------------------------------------------------------

### `bool` assigning

-   Assigning `True` or `False` directly:

In [17]:
my_bool3 = True
print("True is",my_bool3)

# A very bad choice of variable name
true = False
print("False is", true)

True is True
False is False

In [18]:
my_bool1 = 1 == 3
print("1 == 3 is", my_bool1)

# Assigning a more complex Boolean variable
my_bool2 = 1 > 3 or 5 > 3 and not 4 > 5
my_bool2 = (1 > 3) or (5 > 3) and not (4 > 5)
print("(1 > 3) or (5 > 3) and not (4 > 5) is", my_bool2)

1 == 3 is False
(1 > 3) or (5 > 3) and not (4 > 5) is True

## Strings

-   Python uses strings (`str`) for text.
-   Strings can be defined using either double or single quotation
    marks:

In [19]:
my_string1 = "string with double quotation marks"
my_string2 = 'same string but with single marks'
print(my_string1)
print(my_string2)

string with double quotation marks
same string but with single marks

-   Nested strings use both types of quotation marks:

In [20]:
my_string3 = "string with 'single quotation marks'"
my_string4 = 'same string but with "double marks"'
print(my_string3)
print(my_string4)

string with 'single quotation marks'
same string but with "double marks"

------------------------------------------------------------------------

### Long strings and concatenation

-   Long strings can be defined using triple quotation marks:

In [21]:
my_string5 = '''This is a very long string
spanning many lines\nand an extra line break.'''
print(my_string5)

This is a very long string
spanning many lines
and an extra line break.

-   Strings can be concatenated in different ways:

In [22]:
my_string6 = "The answer is " + "42."
my_string7 = "String-1" "String-2"
print(my_string6)
print(my_string7)

The answer is 42.
String-1String-2

------------------------------------------------------------------------

### String conversion

-   Converting objects to `string`:

In [23]:
my_string8 = "The answer is " + str(99.8134) + "."
my_string9 = "The answer is " + str(42) + "."
print(my_string8)
print(my_string9)

The answer is 99.8134.
The answer is 42.

-   Converting `string` to other objects:

In [24]:
my_number = "12.34"
my_float = float(my_number)
print("my_float =",my_float)
print("Data type of my_float is ", type(my_float))

my_float = 12.34
Data type of my_float is  <class 'float'>

------------------------------------------------------------------------

### String replacement

-   Useful when displaying variables:

In [25]:
my_int = 42
my_fl = 84.759
my_str = "message"

# Replacement
my_string1 = f'Have {my_int} and {my_fl} and a {my_str}.'
print(my_string1)

Have 42 and 84.759 and a message.

-   String replacement with fixed width:

In [26]:
print(f'Have {my_int:8} and a {my_str:9}.')
print(f'Have {my_fl:8.2f}.' )

Have       42 and a message  .
Have    84.76.

# Basic data structures

-   A *data structure* is an object that contains other objects.
-   Python has many built-in data structures.
    -   `tuple` and `list` are the most relevant for us.
    -   Other types are `dict` to store objects by key, and `set` to
        collect unique objects.

## Tuples

-   A `tuple` is an ordered collection of objects.
-   Created with (or without) round brackets (`()`):

In [27]:
my_tuple1 = (1, "hello", 5.6)
my_tuple2 = 1, "hello", 5.6
print(my_tuple2)

(1, 'hello', 5.6)

-   Items in tuples can be accessed similar to lists (see below).
-   Tuples are *immutable*: they cannot be changed easily.
-   Lists are more useful in most circumstances.

## Lists and ranges

-   The items in a `list` can be of any type (including a list).

-   Created using square brackets (`[]`):

In [28]:
my_list = [1, "hello", 5.6, -3.9, True, 'end item']
print("my_list =", my_list)
print("The length of the list is", len(my_list))

my_list = [1, 'hello', 5.6, -3.9, True, 'end item']
The length of the list is 6

-   Lists are very powerful; see the Python documentation (The Python
    Software Foundation (2023c)).

------------------------------------------------------------------------

### Accessing list items

-   Python uses *zero-based numbering*: the first item in a list is at
    position 0, the second is at position 1, etc.

In [29]:
print("my_list =",my_list)

my_list = [1, 'hello', 5.6, -3.9, True, 'end item']

In [30]:
print("The first item in the list is", my_list[0])
print("The second item in the list is", my_list[1])
print("The fourth item in the list is", my_list[3])

The first item in the list is 1
The second item in the list is hello
The fourth item in the list is -3.9

-   List items can be accessed from the end of the list by using
    negative indices. The final item is -1, etc.

In [31]:
print("The last item in the list is", my_list[-1])
print("The second to last item is", my_list[-2])

The last item in the list is end item
The second to last item is True

------------------------------------------------------------------------

### List slicing

-   *Slicing* can be used to create a sub-list from an existing list.

In [32]:
my_list = [1, 'hello', 5.6, -3.9, True, 'end item']
print("my_list[2:4] =", my_list[2:4])
print("my_list[:3] =", my_list[:3])
print("my_list[-3:] =", my_list[-3:])

my_list[2:4] = [5.6, -3.9]
my_list[:3] = [1, 'hello', 5.6]
my_list[-3:] = [-3.9, True, 'end item']

-   Lists can be modified by slicing:

In [33]:
my_list[1:4] = ["hello", "greetings"]
print("After replacing a slice:\n\t",my_list)

my_list[2:] = []
print("After removing a slice:",my_list)

After replacing a slice:
     [1, 'hello', 'greetings', True, 'end item']
After removing a slice: [1, 'hello']

------------------------------------------------------------------------

### Reassigning list items

In [34]:
my_list = [1.1, 2.5, -3.7]

my_list[1] = "goodbye"
print("After reassigning item 1:", my_list)

my_list[2] = ['a','2']
print("After reassigning item 2:", my_list)

After reassigning item 1: [1.1, 'goodbye', -3.7]
After reassigning item 2: [1.1, 'goodbye', ['a', '2']]

------------------------------------------------------------------------

### Inserting, deleting and appending list items

-   These operations are time-consuming, use them wisely.
-   Inserting an item:

In [35]:
my_list =  [1.1, 'goodbye', ['a', '2']]
my_list.insert(2, 800)
print("After inserting 800 at position 2:\n\t",my_list)

After inserting 800 at position 2:
     [1.1, 'goodbye', 800, ['a', '2']]

-   Removing item from list:

In [36]:
item2 = my_list.pop(2)
print("Removed item:",item2)
print("List after item removed:",my_list)

Removed item: 800
List after item removed: [1.1, 'goodbye', ['a', '2']]

-   Append item to list:

In [37]:
my_list.append("new end item")
print("After appending item:\n\t",my_list)

After appending item:
     [1.1, 'goodbye', ['a', '2'], 'new end item']

------------------------------------------------------------------------

### Creating lists

-   Creating lists by appending:

In [43]:
new_list1 = []
new_list1.append(4)
new_list1.append(5)
print(new_list1)

[4, 5]

-   Creating lists by conversion:

In [44]:
new_list2 = list("Hello!")
print(new_list2)

['H', 'e', 'l', 'l', 'o', '!']

------------------------------------------------------------------------

### `range` objects

-   `range` is used to create sequences of equispaced integers.

-   Syntax is:

``` {python}
range(start, stop, step)
```

-   If `start` is omitted, it defaults to 0.
-   It `step` is omitted, it defaults to 1. Negative steps are allowed.
-   `stop` is not included in the range.

------------------------------------------------------------------------

### `range` examples

In [45]:
# Numbers 0,...,9
list1 = list(range(10))
print("Integers from 0 to 9:", list1)

Integers from 0 to 9: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

In [46]:
# Numbers 0,2,4,...,10
list2 = list(range(0,12,2))
print("Multiples of 2 from 0 to 10:", list2)

Multiples of 2 from 0 to 10: [0, 2, 4, 6, 8, 10]

In [47]:
# Numbers 0, -4, ..., -10
list3 = list(range(10,-2,-2))
print("Multiples of 2 from 10 to 0:", list3)

Multiples of 2 from 10 to 0: [10, 8, 6, 4, 2, 0]

# Control flow

## Indentation

**Indentation** is very important in Python:

In [48]:
x = 2
print ("x =", x)
if x % 2 == 0:
    print ("x is even")
    print ("to repeat, x is even")
else:
    print ("x is odd")
print ("2x =", 2*x)

x = 2
x is even
to repeat, x is even
2x = 4

## Indentation rules

-   A colon (`:`) denotes the start of an indented block.

-   All code in an indented block must be indented by the same amount
    until the end of the block. The block continues as long as the
    indentation is maintained. Blank lines can be used within indented
    blocks.

-   Google Colab (and many code editors) automatically convert every tab
    stop to two or four spaces, which is recommended as a good
    compromise. However Python doesn’t insist on a specific amount of
    indentation.

## `if`-`else` statements

-   Basic structure:

``` python
if <logical expression a>:
    <code block to execute when a is True>
elif <logical expression b>:
    <code block to execute when a is False and b is True>
elif <logical expression c>:
    <code block to execute when a and b is False and c is True>
...
else:
    <code block executed when all logical expressions above are False>
```

-   The logical expressions are the values of Boolean variables.

------------------------------------------------------------------------

### `if`-`else` examples

In [49]:
y = 7
if y > 0:
    print (y, "is positive.")

7 is positive.

In [50]:
if y % 2 == 0:
    print (y, "is even.")
else:
    print (y, "is not even.")

7 is not even.

In [51]:
y = 7
if 0 < y and y <= 5:
    print ("0 <", y, "<= 5")
elif 5 < y <= 10:
    print ("5 <", y, "<= 10")
else:
    print (y, "> 10 or", y, "<= 0.")

5 < 7 <= 10

------------------------------------------------------------------------

### Conditional expressions

-   Python has a *ternary operator* that allow us to compress
    `if`-`else` statements:

``` python
<expression-if-true> if <condition> else <expression-if-false>
```

A few examples:

In [52]:
z = -3
positive = True if z > 0 else False
print("z is positive:", positive)
absz = z if z > 0 else -z
print("Absolute value of z is", absz)

z is positive: False
Absolute value of z is 3

## `for` loops

-   Useful when the number of repetitions are known.
-   Loops over a `range`, `list` or list-like object.
-   Usage:

``` python
for <variable> in <list or range>:
    <code block to repeat>
```

-   Examples:

In [53]:
# Summing the squares of 0, 1, 2, 3, 4
sum = 0
for item in range(5):
    sum += item**2
print(sum)

30

------------------------------------------------------------------------

### ‘for’ loop examples

In [54]:
# Printing items in a list
my_list = ["first item", "middle item", "last item"]
for item in my_list:
    print(item)

first item
middle item
last item

In [55]:
# More complex example with nested indentation
for i in range(1, 5):
    if i % 2 == 0:
        message = f"{i} is an even number."
    else:
        message = f"{i} is an odd number."
    print(message)

1 is an odd number.
2 is an even number.
3 is an odd number.
4 is an even number.

------------------------------------------------------------------------

### List comprehensions

-   `for` and `if` can be used to filter lists:

``` python
[<expression> for <object> in <list/collection> if <logical condition>]
```

-   Examples:

In [56]:
my_list = [0, 1, 2, 3, 4, 5]
my_list_squares = [item**2 for item in my_list]
print("list of squares =", my_list_squares)
my_list_even = [i for i in my_list if i % 2 == 0]
print("list of even numbers =", my_list_even)

list of squares = [0, 1, 4, 9, 16, 25]
list of even numbers = [0, 2, 4]

------------------------------------------------------------------------

## `while` loops

-   Useful when code is repeated while a logical condition remains true.
-   Useful when unsure how many repetitions are needed.

``` python
while <logical expression>:
    <code block to execute repeatedly while logical expression remains True>
```

-   Example:

In [57]:
sum = 0
item = 0
while item < 5:
    sum += item**2
    item += 1
print(sum)

30

------------------------------------------------------------------------

### `while` examples

-   Find the smallest power of 2 larger than 1000:

In [58]:
p = 0
while 2**p <= 1000:
    p += 1
print(f"p = {p} and " 
      f"{2**(p-1)} = 2^{p-1} <= 1000 < 2^{p} = {2**p}")

p = 10 and 512 = 2^9 <= 1000 < 2^10 = 1024

-   If the logical expression is `False` from the start, then the
    `while` loop is never entered:

In [59]:
p = 0
while 2**p <= 0:
    p += 1
print(f"p = {p}, 2^{p-1} = {2**(p-1)}, 2^{p} = {2**p}")

p = 0, 2^-1 = 0.5, 2^0 = 1

# Functions

-   Functions are useful for:
    -   Reusing rather than repeating code we’ve already written.
    -   Separating different tasks in a program.
    -   Structuring our code by grouping tasks together.

## Defining a function

-   Syntax:

``` python
def <function name> (<argument(s)>):
    """<optional string literal 
    with brief description>"""
    <code block>
    return <value(s)>
```

-   Function names follow the same rules as variable names.

-   `return` statement is optional:

    -   Special value `None` is returned if `return` statement is
        missing.
    -   Multiple `return` statements are allowed.

## Function arguments

-   Arguments are separated by commas (`,`).
-   **Positional arguments** must be specified (in correct order) to use
    a function.
-   **Keyword arguments** are used to specify optional arguments and/or
    default values. Can be specified in any order.
-   Positional arguments listed first, before keyword arguments.
-   Example:

In [60]:
def is_multiple (number, factor = 2):
    "Tests whether one integer is a multiple of another."
    if number % factor == 0:
        return True
    else:
        return False

print("6 is a multiple of 2:", is_multiple(6))
print("6 is a multiple of 3:", is_multiple (6, factor=3))

6 is a multiple of 2: True
6 is a multiple of 3: True

## Function docstring

-   Comment string helps make code readable and helps with coding.
-   Optional but strongly recommended (omitted on slides for fit only!).
-   Example:

In [61]:
help(is_multiple)

Help on function is_multiple in module __main__:

is_multiple(number, factor=2)
    Tests whether one integer is a multiple of another.


## Local and global namespace

-   Variables created inside the function are assigned in local
    *namespace* and destroyed when the function exits.
-   Example:

``` python
def my_function():
    a = 6
    print("a in function =",a)
    
my_function()
print("a outside function =",a)
```

a in function = 6
NameError: name 'a' is not defined

------------------------------------------------------------------------

### Higher scope

-   Functions can also access variables created outside the function in
    higher (or *global*) *scope*.
-   Useful but dangerous because:
    -   Global variables can change unexpectedly.
    -   It can be hard to keep track of which functions use global
        variables.
-   Example:

In [63]:
a = 6

def my_function():
    print("a =",a)
    
my_function()

a = 6

------------------------------------------------------------------------

### Global variables

-   The value of a global variable cannot be changed inside a function
    (without using the `global` keyword).
-   Example:

In [64]:
a = 6

def my_function():
    a = 5
    print("inside function a =",a)
    
my_function()
print("outside function a =",a)

inside function a = 5
outside function a = 6

------------------------------------------------------------------------

### Use arguments instead

-   Much better to use arguments instead of global variables.
-   Example:

In [65]:
a = 6

def my_function(a):
    print("inside function a =",a)
    
my_function(a)
my_function(5)

inside function a = 6
inside function a = 5

## Returning multiple objects

-   It’s possible to return multiple items from a function.
-   The function is returning an object of type `tuple`, which is
    unpacked into result objects if needed.
-   Example:

In [66]:
def my_function():
    return 1, 2, 3

a, b, c = my_function()
print (f"a = {a}, b = {b}, c = {c}")

result = my_function()
print (f"result = {result} can be unpacked into {result[0]} and {result[1]} and {result[2]}")

a = 1, b = 2, c = 3
result = (1, 2, 3) can be unpacked into 1 and 2 and 3

## Anonymous (`lambda`) functions

-   An anonymous or lambda function is a way of writing a function
    consisting of a single statement.
-   Syntax:

``` python
lambda <argument>: <value>
```

-   Equivalent ways of defining the same function:

In [67]:
def my_function(x):
    return x + 2

my_function = lambda x: x + 2

## Functions are objects

-   Functions are objects and can be passed as arguments.
-   `lambda` functions are very useful for this.
-   Example:

In [68]:
def apply_to_list (my_list, my_func):
    return [my_func(x) for x in my_list]

result1 = apply_to_list([1,2,3], my_function)
result2 = apply_to_list([1,2,3], lambda x: x-2)
print(result1)
print(result2)

[3, 4, 5]
[-1, 0, 1]

-   Functions can also be returned from functions.

# Further reading

## Books

-   These slides are based on pp. 62–80 by Hilpisch (2019).

-   A few recommended books:

    -   Smith (2020), Sections 3.2–3.5.

    -   McKinney (2022), Sections 2.3 and 3.1–3.2.

    -   Downey (2016), Chapters 2, 5, 7, 8, 10, 12.

## Online resources

-   There are many good resources for learning Python programming.

-   Check that the resource covers Python 3—it is very different from
    Python 2. One clue is the `print` statement—Python 2 has `print a`
    but Python 3 has `print(a)`.

-   A few recommended online tutorials:

    -   Bostroem, Bekolay, and Staneva (2016)
    -   Pussinen et al. (2023)
    -   McOwen (2021)
    -   The Python Software Foundation (2024)

## References

Bostroem, Azalee, Trevor Bekolay, and Valentina Staneva. 2016. “Software
Carpentry: Programming with Python.”
<https://doi.org/10.5281/zenodo.57492>.

Downey, Allen. 2016. *Think Python: How to Think Like a Computer
Scientist*. 2nd ed. O’Reilly.
<https://greenteapress.com/wp/think-python-2e/>.

Hilpisch, Yves. 2019. *Python for Finance: Mastering Data-Driven
Finance*. 2nd ed. O’Reilly.

IEEE. 2008. “IEEE Standard for Floating-Point Arithmetic.”
<https://doi.org/10.1109/IEEESTD.2008.4610935>.

McKinney, Wes. 2022. *Python for Data Analysis: Data Wrangling with
Pandas, NumPy & Jupyter*. 3rd edition. O’Reilly.
<https://wesmckinney.com/book/>.

McOwen, Sean. 2021. “FinanceAndPython.com Tutorial: Python Primer.”
<https://github.com/FinanceAndPython/FinanceAndPython.com/tree/master/Python%20Primer>.

Pussinen, Jerry et al. 2023. “Learn Python 3.”
<https://github.com/jerry-git/learn-python3>.

Smith, Einar. 2020. *Introduction to the Tools of Scientific Computing*.
Springer.
<https://yorsearch.york.ac.uk/permalink/f/7htm32/TN_cdi_openaire_primary_doi_dedup_9c7df322731b6cea4969322abe566638>.

The Python Software Foundation. 2023a.
“<span class="nocase">decimal</span>: Decimal Fixed Point and Floating
Point Arithmetic.” <https://docs.python.org/3/library/decimal.html>.

———. 2023b. “<span class="nocase">keyword</span>: Testing for Python
Keywords.” <https://docs.python.org/3/library/keyword.html>.

———. 2023c. “Sequence Types: <span class="nocase">list</span>,
<span class="nocase">tuple</span>, <span class="nocase">range</span>.”
<https://docs.python.org/3/library/stdtypes.html#text-sequence-type-str>.

———. 2024. “The Python Tutorial.”
<https://docs.python.org/3/tutorial/index.html>.